src/compiler/sourcemap.ts

```diff
@@ -322,44 +489,205 @@ namespace ts {
+        function emitTokenStart(token: SyntaxKind, tokenStartPos: number, contextNode?: Node, ignoreTokenCallback?: (node: Node, token: SyntaxKind) => boolean, getTokenTextRangeCallback?: (node: Node, token: SyntaxKind) => TextRange): number {
+            if (contextNode) {
+                if (ignoreTokenCallback(contextNode, token)) {
+                    return skipTrivia(currentSourceText, tokenStartPos);
+                }
+
+                const range = getTokenTextRangeCallback(contextNode, token);
+                if (range) {
+                    tokenStartPos = range.pos;
+                }
+            }
+
+            tokenStartPos = skipTrivia(currentSourceText, tokenStartPos);
+            emitPos(tokenStartPos);
+            return tokenStartPos;
+        }
+
+        /**
+         * Emits a mapping for the end position of a token.
+         *
+         * If the token's end position is synthetic (undefined or a negative value), no mapping
+         * will be created.
+         *
+         * @param token The token to emit.
+         * @param tokenEndPos The end position of the token.
+         * @returns The end position of the token.
+         */
+        function emitTokenEnd(token: SyntaxKind, tokenEndPos: number): number;
+        /**
+         * Emits a mapping for the end position of a token.
+         *
+         * If the token's end position is synthetic (undefined or a negative value), no mapping
+         * will be created.
+         *
+         * @param token The token to emit.
+         * @param tokenEndPos The end position of the token.
+         * @param contextNode The node containing this token.
+         * @param ignoreTokenCallback A callback used to determine whether to skip source map
+         *        emit for the end position of this token.
+         * @param getTokenTextRangeCallback A callback used to get a custom source
+         *        map range for this node.
+         * @returns The end position of the token.
+         */
+        function emitTokenEnd(token: SyntaxKind, tokenEndPos: number, contextNode: Node, ignoreTokenCallback: (node: Node, token: SyntaxKind) => boolean, getTokenTextRangeCallback: (node: Node, token: SyntaxKind) => TextRange): number;
+        function emitTokenEnd(token: SyntaxKind, tokenEndPos: number, contextNode?: Node, ignoreTokenCallback?: (node: Node, token: SyntaxKind) => boolean, getTokenTextRangeCallback?: (node: Node, token: SyntaxKind) => TextRange): number {
+            if (contextNode) {
+                if (ignoreTokenCallback(contextNode, token)) {
+                    return tokenEndPos;
+                }
+
+                const range = getTokenTextRangeCallback(contextNode, token);
+                if (range) {
+                    tokenEndPos = range.end;
+                }
+            }
+
+            emitPos(tokenEndPos);
+            return tokenEndPos;
+        }
+
+
+        // @deprecated

```

Comment:
It is used by the old emitter, and will be removed when the old emitter is removed.